src/radical/pilot/compute_pilot.py

```diff
@@ -137,12 +137,16 @@ def __str__(self):
 
     # --------------------------------------------------------------------------
     #
-    def _default_state_cb(self, pilot, state):
+    def _default_state_cb(self, pilot, state=None):
 
-        self._log.info("[Callback]: pilot %s state: %s.", self.uid, self.state)
 
-        if self.state == rps.FAILED and self._exit_on_error:
-            self._log.error("[Callback]: pilot '%s' failed - exit", self.uid)
+        uid   = self.uid
+        state = self.state
+
+        self._log.info("[Callback]: pilot %s state: %s.", uid, state)
+
+        if state == rps.FAILED and self._exit_on_error:
+            self._log.error("[Callback]: pilot '%s' failed (exit)", uid)
 
             # There are different ways to tell main...
```

Comment:
Is this still the case or do we have just one way now?

Reply:
The other ways are commented out below.  All of them have different trade-offs, that's and I want to keep the different options around in case we need to switch.  The current version is very reliable, but leads to a confusing `KeyboardInterrupt`, and uses an undocumented API call.  `raise` is unreliable, `os.kill` is very reliable, but kills the parent thread w/o warning or trace, `sys.exit` would be perfect, but internally uses signals which are unreliable in threaded applications.